go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -212,18 +351,81 @@ func (npw *nodePortWatcher) SyncServices(services []interface{}) {
 				serviceInterface)
 			continue
 		}
-		npw.updateServiceFlowCache(service, true)
+		npw.updateServiceFlowCache(service, true, false)
 	}
 
 	npw.ofm.requestFlowSync()
-	syncSharedGatewayIptRules(services)
+	syncSharedGatewayIptRules(services, false, npw.nodeIPManager.addresses.List())
+}
+
+// For endpoint Event functions and exernalTrafficPolicy:local
+//     An action only ever occurs (i.e epHostLocal=True) if
+//			1. The Service Exists &&
+// 			2. Svc externalTrafficpolicy:local &&
+//			3. Svc Has host networked Backends
+//      Otherwise the endpoint event is ignored
+
+// Add new Local Traffic flows if endpoints are local
+func (npw *nodePortWatcher) AddEndpoints(ep *kapi.Endpoints) {
+	var svc interface{}
+	var err error
+	name := ktypes.NamespacedName{Namespace: ep.Namespace, Name: ep.Name}
+
+	// Endpoint add Event can arrive before service add event, get and create it if it doesn't exist
+	svc, exists := npw.services.Load(name)
+	if !exists {
```

Comment:
This isn't thread safe because after you have loaded, and found it doesnt exist, now the service thread could have gotten the event and create it. You have to use LoadorStore to ensure the operation is atomic and add service handler wont interfere.

Reply:
But we don't have the service yet..

Reply:
I should have fixed this, basically the AddService and AddEndpoint handlers race to see who can win and then handles each case accordingly